core/frontend/src/components/autopilot/FirmwareManager.vue

```diff
@@ -0,0 +1,269 @@
+      return this.upload_type === UploadType.Cloud ? this.cloud_firmware_available : this.file_firmware_available
+    },
+  },
+  methods: {
+    async updateAvailableFirmwares(): Promise<void> {
+      await axios({
+        method: 'get',
+        url: `${autopilot_store.API_URL}/available_firmwares`,
+        timeout: 2000,
+        params: { vehicle: this.chosen_vehicle },
+      })
+        .then((response) => {
+          this.available_firmwares = response.data
+        })
+        .catch((error) => {
+          notification_store.pushNotification(new LiveNotification(
+            NotificationLevel.Error,
+            autopilot_service,
+            'FIRMWARE_FETCH_FAIL',
+            `Could not fetch available firmwares: ${error.message}.`,
+          ))
+        })
+    },
+    setCloudFirmwareAvailable(): void {
+      this.cloud_firmware_available = true
+    },
+    setFileFirmwareAvailable(file: Blob): void {
+      this.firmware_file = file
+      this.file_firmware_available = true
+    },
+    async installFirmware(): Promise<void> {
+      this.installing = true
+      this.install_completed = false
+      this.successfull_install = false
+      const axios_request_config: AxiosRequestConfig = {
+        method: 'post',
+        timeout: 30000,
+      }
+      if (this.upload_type === UploadType.Cloud) {
+        // Populate request with data for cloud install
+        Object.assign(axios_request_config, {
+          url: `${autopilot_store.API_URL}/install_firmware_from_url`,
+          params: { url: this.chosen_firmware_url },
+        })
+      } else {
+        // Populate request with data for file install
+        if (!this.firmware_file) {
+          notification_store.pushNotification(new LiveNotification(
+            NotificationLevel.Warning,
+            autopilot_service,
+            'FILE_FIRMWARE_UPLOAD_FAIL',
+            'Could not upload firmware: no firmware file selected.',
+          ))
+          return
+        }
+        const form_data = new FormData()
+        form_data.append('binary', this.firmware_file)
+        Object.assign(axios_request_config, {
+          url: `${autopilot_store.API_URL}/install_firmware_from_file`,
+          headers: { 'Content-Type': 'multipart/form-data' },
+          data: form_data,
+        })
+      }
+
+      await axios(axios_request_config)
+        .then(() => {
+          this.successfull_install = true
+          this.install_result_message = 'Successfully installed new firmware'
+        })
+        .catch((error) => {
+          notification_store.pushNotification(new LiveNotification(
+            NotificationLevel.Error,
+            autopilot_service,
+            'FILE_FIRMWARE_UPLOAD_FAIL',
+            `Could not upload firmware: ${error}.`,
+          ))
+          this.successfull_install = false
+          try {
+            this.install_result_message = error.response.data.message
+          } catch {
+            this.install_result_message = ''
```

Comment:
This should contains something like "Invalid backend error message."

Reply:
Good one.